aten/src/THC/THCAtomics.cuh

```diff
@@ -247,4 +281,57 @@ static inline __device__ void atomicAdd(bool *address, bool val) {
   gpuAtomicAdd(address, val);
 }
 
+// Atomic multiplication implementation.
+
+// This needs to exist since we use GPU atomics in the scatter/gather
+// functionality and it cannot be specialized for only certain types
+// since it must handle things like assignment operations for generic
+// tensors.
+template<typename T>
+inline __device__ void gpuAtomicMul(T * address, T val) {
+  assert(0);
+}
+
+inline __device__ at::Half gpuAtomicMul(at::Half * address, at::Half val) {
+  return AtomicFPOp<at::Half>()(address, val,
+                                [](at::Half bsum, at::Half val) {
+                                  return THCNumerics<at::Half>::mul(bsum, val);
+                                });
+}
+
+inline __device__ at::BFloat16 gpuAtomicMul(at::BFloat16 * address, at::BFloat16 val) {
+  return AtomicFPOp<at::BFloat16>()(address, val,
+                                    [](at::BFloat16 bsum, at::BFloat16 val) {
+                                      return THCNumerics<at::BFloat16>::mul(bsum, val);
+                                    });    
+}
+    
+inline __device__ double gpuAtomicMul(double * address, double val) {
+  return AtomicFPOp<double>()(address, val,
+                              [](double val, unsigned long long int assumed) {
+                                return __double_as_longlong(val * __longlong_as_double(assumed));
+                              });
+}
+
+// Dont use a templated function for this since the addition function defaults to the CUDA built-in.
+inline __device__ float gpuAtomicMul (float * address, float val) {
+  unsigned int* address_as_ull = (unsigned int*)address;
+  unsigned int old = *address_as_ull;
+  unsigned int assumed;
+
+  do {
+    assumed = old;
+    old = atomicCAS(address_as_ull, assumed,
+                    __float_as_int(val *
+                                   __int_as_float(assumed)));
+
+    // Note: uses integer comparison to avoid hang in case of NaN (since NaN != NaN)
+  } while (assumed != old);
+}
+
+template<typename T>
+inline __device__ void gpuAtomicMul (c10::complex<T> * address, c10::complex<T> val) {
```

Comment:
This is not complex multiplication and properly enabling complex atomic multiplication is very challenging. I suggest not enabling scatter_reduce with mul for complex (and same for future work, we don't enable complex min and max otherwise, so we should not enable them for scatter operations). 
As an aside, complex gpuAtomicAdd as it exists now it also not atomic, so should not be called that, but no action required for this PR.

Reply:
Ah my bad. Should not missed this one sorry. Will change it.